E2_ESRGAN/lib/dataset.py

```diff
@@ -1,44 +1,101 @@
-    return (low_resolution, high_resolution)
+        high_resolution,
+        [dimension // factor, dimension // factor],
+        method=method)
+    low_resolution = tf.clip_by_value(low_resolution, 0, 255)
+    high_resolution = tf.clip_by_value(high_resolution, 0, 255)
+    return low_resolution, high_resolution
   return scale_fn
 
 
+def augment_image(
+        low_res_map_fn,
+        brightness_delta=0.05,
+        contrast_factor=[0.7, 1.3],
+        saturation=[0.6, 1.6]):
+  def augment_fn(_, high_resolution, *args, **kwargs):
+    high_resolution = tf.image.random_flip_left_right(high_resolution)
+    high_resolution = tf.image.random_flip_up_down(high_resolution)
+    high_resolution = tf.image.random_brightness(
+        high_resolution, max_delta=brightness_delta)
+    if contrast_factor:
+      high_resolution = tf.image.random_contrast(
+          high_resolution, *contrast_factor)
+    if saturation:
+      high_resolution = tf.image.random_saturation(
+          high_resolution, *saturation)
+    low_res, high_res = low_res_map_fn(high_resolution)
+    return low_res, high_res
+  return augment_fn
+
+
+def reform_dataset(dataset, types):
+  """ Helper function to convert the output_dtype of the dataset
+      from (tf.float32, tf.uint8) to desired dtype
+  """
+  def generator_fn():
+    for data in dataset:
+      yield data[0], data[1]
+  return tf.data.Dataset.from_generator(
+      generator_fn, types, (tf.TensorShape([None, None, 3]), tf.TensorShape(None)))
+
+
 def load_dataset_directory(
         name,
         directory,
         low_res_map_fn,
         batch_size=32,
         iterations=1,
         shuffle=True,
+        augment=True,
         cache_dir="cache/",
         buffer_size=io.DEFAULT_BUFFER_SIZE):
 
   if not tf.io.gfile.exists(cache_dir):
     tf.io.gfile.mkdir(cache_dir)
   dl_config = tfds.download.DownloadConfig(manual_dir=directory)
-  dataset = (tfds.load("image_label_folder/dataset_name=%s" % name,
-                       split="train",
-                       as_supervised=True,
-                       download_and_prepare_kwargs={"download_config": dl_config})
-             .map(low_res_map_fn, num_parallel_calls=tf.data.experimental.AUTOTUNE)
+  dataset = reform_dataset(
+      tfds.load(
+          "image_label_folder/dataset_name=%s" %
+          name,
+          split="train",
+          as_supervised=True,
+          download_and_prepare_kwargs={
+              "download_config": dl_config}),
+      (tf.float32,
+       tf.float32))
+  dataset = (dataset.map(low_res_map_fn, num_parallel_calls=tf.data.experimental.AUTOTUNE)
              .batch(batch_size)
              .prefetch(buffer_size)
              .cache(cache_dir))
   if shuffle:
     dataset = dataset.shuffle(buffer_size, reshuffle_each_iteration=True)
   # Repeat step should be after shuffle for lower memory footprint
-  dataset = dataset.repeat(iterations)
+  if augment:
+    dataset = dataset.map(
```

Comment:
Hey @srjoglekar246 , I want to call this map function for every iteration. since it produces new set of images with augmentation.
so, putting it after shuffle does it make sense? since shuffle is reshuffling every iteration, I thought may be, the dataset will be be fooled into thinking that each permutation is a different dataset and will map the function, everytime. I'm not really sure about this. can you help out?

Reply:
In spirit, it seems fine to me. However, the augmentation part(`augment_fn`) might be time-consuming, especially since you are applying them to high-res images. I would suggest doing each augmentation randomly (Say with 50% probability?) - and check the effect on training time. And for a small fraction (say 20%), do no augmentations.

Reply:
yea. that sounds logical. It can be fixed, easily I guess.